### PR TITLE
UL-110: Initial commit of documentation for new UL Image API....

### PR DESCRIPTION
**Pull request by the-t-in-rtf**
See [UL-110](https://issues.gpii.net/browse/UL-110) and [UL-108](https://issues.gpii.net/browse/UL-108) for details.

### CODE REVIEW

File: docs/api-rest-docs.md [hunk 1]
```diff
@@ -0,0 +1,281 @@
...
++ Query Parameters
+    + `height` (optional, number) ... The desired height of the image.  If `height` is supplied and ``width` is omitted, the width will be determined using the aspect ratio of the original image.
+    + `width` (optional, number) ... The desired width of the image.  If `width` is supplied and `height` is omitted, the height will be determined using the aspect ratio of the original image.
+
+If both `height` and `width` are specified, the image will be scaled proportionately and mounted on a background with that
+exact size.  See [the Sharp documentation](http://sharp.dimens.io/en/stable/api/#embed) for details.
+
+# `HEAD /api/images/file/:uid/:source/:image_id{?height&width}`
+
+Returns just the HTTP request headers for a given image, including:
```
Comment (amb26) [inline docs/api-rest-docs.md]: etags? What scheme will be used to check staleness?

File: docs/api-client-docs.md [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+A component to edit the metadata associated with a single image file.
+
+# `gpii.ul.images.client.gallery`
+
+A component to display a collection of images as thumbnails, with support for zooming in on individual images.
+
+# `gpii.ul.images.client.gallery.edit`
+
+A component to enable reviewers to choose which images (and in which order) to display in a gallery view.
```
Comment (amb26) [inline docs/api-client-docs.md:28]: You may be interested in the "imageReorderer" and "uploader" components that we worked on in the previous decade. They may not be appropriate, but they have at least had a significant amount of accessibility work invested in them, as well as ensuring that they continue to work with the current Infusion:
https://github.com/fluid-project/infusion/tree/master/demos/reorderer/imageReorderer
https://github.com/fluid-project/infusion/tree/master/src/components/uploader
If they are not appropriate, it would be useful to know what features you would like from them in order to inform any future work.

File: README.md [hunk 3]
```diff
@@ -1 +1,60 @@
...
+## The API
+
+The API provides a series of REST inputs.  For details regarding the endpoints, their accepted input, and their return
+values, see [docs/api-rest-docs.md](the API documentation).
+
+The API is built as a series of [Fluid components](http://fluidproject.org/). For more information, see
+[docs/api-component-docs.md](the API component documentation).  In addition, the API is built using the following key
+libraries:
+
+1. [Sharp](http://sharp.dimens.io/), for image resizing.
```
Comment (amb26) [inline README.md:37]: Somewhere (perhaps not here) there should be materials for a technology evaluation exercise - what requirements did you have, how did you evaluate them, and what alternatives did you consider for these libraries.

File: docs/api-client-docs.md [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+To use them, you will need to:
+ 
+ 1. Provide markup that will hold the component's content.
+ 2. Include Fluid and all client-side dependencies from this package.  See the tests for examples.
+ 3. Make the required templates available using the ["inline" template middleware provided by the `gpii-handlebars` package].
+ 4. Instantiate the component.
+
+// TODO:  Flesh this out with practical examples, including details about working with templates.
+
```
Comment (amb26) [inline docs/api-client-docs.md:13]: Yes, a working sample is worth a thousand instructions

File: README.md [hunk 5]
```diff
@@ -1 +1,60 @@
-# ul-image-api
\ No newline at end of file
+# UL Image API
```
Comment (amb26) [inline README.md:1]: Shouldn't this be partnered with a GPII organisation repository as with all the others?

File: docs/api-rest-docs.md (hunk 1 above)
Comment (the-t-in-rtf) [reply docs/api-rest-docs.md]: For original (unresized) images, I was hoping to defer to the [static content router](https://github.com/the-t-in-rtf/gpii-express/blob/GPII-2161/docs/router.md#gpiiexpressrouterstatic).  Looking through the libraries we wrap, it looks like the staleness is based on the [last modification date](https://github.com/expressjs/serve-static?_ga=1.235046086.618322545.1478786081#lastmodified) of the file.  A challenge of passing through to the static router is ensuring that the path used for the "outer" middleware can be used to locate the filesystem content.  I believe this is surmountable as long as our filesystem storage follows the same uid/source/image_id structure.

For resized images, it depends on whether we save resized versions of each image.  If we resize the first time someone asks for a given size and save to file, we can then use the static content router to serve it up (although we might need to change the path structure to make this possible).   I am hoping to avoid saving resized images, as it has the potential to bloat our storage, and [resizing with sharp is fairly cheap](http://sharp.dimens.io/en/stable/performance/).  It also pushes both the resizing and a file write onto the first user, in practice I suspect that resizing every time will be just fine performance wise.

If we choose not to save resized images, we need a way to respond to HEAD requests without actually resizing the image. Refreshing my memory about useful [etags](https://github.com/expressjs/serve-static?_ga=1.235046086.618322545.1478786081#lastmodified), it sounds like we need to ensure the value is unique for a given image and size.  I would propose combining the image_id, the size parameters, and the last modification date of the original file to form the etag.  Each of these is low-cost to discover, and hence appropriate for a HEAD request.

File: docs/api-client-docs.md (hunk 2 above)
Comment (the-t-in-rtf) [reply docs/api-client-docs.md:28]: I was thinking along the same lines.  I would use the image reorder for "gallery" editing, likely there would be no changes to the underlying component.

I suspect the uploader will require more effort, but that's just an impression.  I am happy to start with the uploader and discuss further as I have clear examples of what works or doesn't work for this use case.

File: README.md (hunk 5 above)
Comment (the-t-in-rtf) [reply README.md:1]: Agreed, I will migrate to using a GPII repo before any code is submitted.  The ticket for this is [GPII-2169](https://issues.gpii.net/browse/GPII-2169).

File: README.md (hunk 3 above)
Comment (the-t-in-rtf) [reply README.md:37]: I kept rough notes that I am happy to flesh out regarding the two key libraries chosen (sharp, and multer).  I will comment on both here, and you can advise as to whether we need to create a wiki page, a mailing list post, etc. for each.

Comment (the-t-in-rtf) [reply README.md:37]: Image resizing was one of the stated goals of the work, as we want to support both "thumbnails" and original images.  My starting assumption was that we needed something that supported common image formats, especially those we already have in our collection of images coming from EASTIN.  I prepared a report to confirm our holdings, we have a mixture of PNG, GIF, and JPEG images, with one webp file.  As webp is not widely supported within browsers, I considered that an optional requirement.

When looking for a library, I did a number of google searches and npm searches, and looked through a range of candidates.  A few names kept coming up, all of which are mentioned in the excellent [performance docs for "sharp"](http://sharp.dimens.io/en/stable/performance/):

1. [sharp](http://sharp.dimens.io/)
2. [lwip](https://www.npmjs.com/package/lwip)
3. [jimp](https://www.npmjs.com/package/jimp)
4. [imagemagick](https://www.npmjs.com/package/imagemagick-native)

Of the pure javascript approaches, sharp was the fastest by far and gave more of an impression of maturity as a project/community than "jimp" or "lwip".  By that I mean that the docs were more complete, the community was a healthy size, the release cadence seemed healthy, and the API seemed well thought out.  Imagemagick is a classic that I used productively in the late 90s, but it is slower, less well maintained, and would increase the deployment and test complexity without adding any real value.

In summary, sharp is by far the fastest option I could find, and my impression was that it was the best starting point for image resizing.

Comment (the-t-in-rtf) [reply README.md:37]: The other key piece of technology is something that handles multipart/form uploads.  It needs to:

1. Accept file content encoded using multipart/form encoding.
2. Refuse to accept overly large files.
3. Refuse to read an overly large amount of field data into memory.
4. Refuse to accept unacceptable file types.

I started with the [express documentation](https://github.com/expressjs/body-parser), which mentions a range of candidates.  Of those, I almost immediately gravitated towards multer because I could see clearly from the docs that they at least intend for people to have a lot of control over the way it works from the outside.  For example, check out  [the amount of control multer gives when choosing what content to accept](https://www.npmjs.com/package/multer#limits).  You can either use declarative options to configure it, or if that's too limiting, you can [wire in your own filter function](https://www.npmjs.com/package/multer#filefilter).  They also have clear docs on replacing the storage engine, which appealed to me as it gives us the option to explore using versioned storage down the road.

At least from their docs, all of the rest (multipart, busboy, formidable) seemed to offer only fixed configuration options.  They also do not allow you to reject files of a particular type.

In summary, multer was the only candidate that met all of our needs.  It also seems to give us the most room to expand or extend its behavior in the future.

File: docs/api-client-docs.md (hunk 4 above)
Comment (the-t-in-rtf) [reply docs/api-client-docs.md:13]: I appreciate your indulging my "documentation driven design", working samples will be forthcoming.... :)

File: docs/api-rest-docs.md (hunk 1 above)
Comment (amb26) [reply docs/api-rest-docs.md]: Thanks - I think this info could helpfully go into a "Task" JIRA issue named something like "Implement strategy for reporting staleness for image resources" or so

File: README.md (hunk 3 above)
Comment (amb26) [reply README.md:37]: Cheers - looks like an appropriate place for this info is on a wiki page linked from the following head: https://wiki.gpii.net/w/Technology_Evaluation

Comment (the-t-in-rtf) [reply README.md:37]: Is there a template?  Also, are there other examples? I remember another one Alan did as well, but don't see that.

File: docs/api-rest-docs.md (hunk 1 above)
Comment (the-t-in-rtf) [reply docs/api-rest-docs.md]: Done, see [UL-112](https://issues.gpii.net/browse/UL-112).

File: README.md (hunk 3 above)
Comment (the-t-in-rtf) [reply README.md:37]: For anyone else looking for this, I backtracked through my vague recollections of previous discussions and eventually found this thread, which includes links to the governance docs and examples of previous documents:

http://lists.idrc.ocad.ca/pipermail/architecture/2016-March/003946.html

Comment (the-t-in-rtf) [reply README.md:37]: See https://wiki.gpii.net/w/Technology_Evaluation for both writeups.

**Comment by the-t-in-rtf** [comment]
@amb26, I'd like to close this out, we can continue the review once there is actual code in the monorepo.

**Comment by amb26** [comment]
Given this has been issued against one of your own repos, I can't merge it myself but I'm happy to sign it off on the basis of "looks reasonable to me", which sadly can't equate to much assurance of review quality - the API does look reasonable as far as it goes but it's hard to give such a thing much assessment in advance of there being a working artefact. If I were an IBM employee, perhaps things would be different :)
